synapse/storage/databases/main/event_federation.py

```diff
@@ -588,6 +610,122 @@ def _get_auth_chain_difference_using_cover_index_txn(
 
         return result
 
+    def _fixup_auth_chain_difference_sets(
+        self,
+        txn: LoggingTransaction,
+        room_id: str,
+        state_sets: List[Set[str]],
+        events_missing_chain_info: Set[str],
+        events_that_have_chain_index: Collection[str],
+    ) -> Set[str]:
+        """Helper for `_get_auth_chain_difference_using_cover_index_txn` to
+        handle the case where we haven't calculated the chain cover index for
+        all events.
+
+        This modifies `state_sets` so that they only include events that have a
+        chain cover index, and returns a set of event IDs that are part of the
+        auth difference.
```

Comment:
ah, here's the hidden `mut` keyword :p

Took me a moment to notice that modifying `state_sets` is part of the result but not sure I have a better suggestion

Reply:
It's icky, but here we are. I mostly wanted to avoid needlessly copying the state sets when its easier to just mutate tehm

Reply:
yeah, reasonable.